### DISCUSSION TITLE
How to test async useEffect?

### DISCUSSION THREAD

**Discussion by opsb** [GregoryConrad/rearch-dart#182] [Q&A]
Given the following

```dart
Report salesReport(CapsuleHandle use) {
  final (timeRange, _) = use(timeRangeCapsule);
  final (report, setReport) = use.state(initialReport);
  final api = use(apiCapsule);

  updateReport() async {
    final updated = api.fetchReport(timeRange);
    setReport(updated);
  }

  use.effect(() {
    updateReport();
  }, [timeRange])

  return report;
}
```

How could I test that the report is updated when the timeRangeCapsule changes? More specifically, how should you wait in the test for the `api.fetchReport` to complete?

**Comment by GregoryConrad** [comment]
What you have is looking decent so far!

instead of use.effect and use.state, create a future with use.memo, and then use.future with that created future (or just return the future directly instead of an AsyncValue, depending on your needs). That way you don’t need to manually set the state.

away from my computer so I can’t give you a proper typed-out example right now, but does that make sense?

**Reply by GregoryConrad** [reply]
Btw, you may just want to return a future directly because then you can await that future in your test for simplicity. If you use.future, you’d have to container.warmUp on the capsule (which would work, but maybe a little more effort)

**Reply by GregoryConrad** [reply]
Also, looking at this again: you probably don’t need a use.memo either. Just:

`Future<Report> salesReportAsyncCapsule(CapsuleHandle use) => use(apiCapsule).fetchReport(use(timeRangeCapsule));`

And if you need it you can make a capsule that wraps around that one and does a use.future. To add in the initial data, you can do AsyncValue.copyWithPrevious (or whatever I called the function)

**Reply by opsb** [reply]
For the use case I have in mind I need the previous value until the new one has loaded. I can use `use.previous` to achieve that, just playing around to see what's the best balance between testable and ergonomic.

**Reply by GregoryConrad** [reply]
Alright, so make a new capsule around the one that returns a future:

`use.future(use(salesReportAsyncCapsule)).copyWithPrevious(initialReport);`

then you can use one of the AsyncValue extension methods and rest assured that there will always be some data in it, even in loading/error cases

**Reply by GregoryConrad** [reply]
Sorry, here’s the function I was thinking of:
https://github.com/GregoryConrad/rearch-dart/blob/238507a3d0a76a29be9876b71fdb0c4699e8d78e/packages/rearch/lib/src/types.dart#L301

**Reply by opsb** [reply]
That's nice, very neat

**Comment by opsb** [comment]
This is a more complicated example. There's a shared buffer which is updated using both a stream and a fetch. A stream could be used to combine them (the codebase currently uses RxDart) and then do less work within Rearch but I'm curious how you would approach making this scenario testable with a Rearch based implementation.

```dart
  IOrderedSet<GlucoseReading> glucoseReadingsBuffer(CapsuleHandle use) {
    final (timeFrame, _) = use(timeFrameCapsule);
    final (buffer, setBuffer) = use.state(IOrderedSet<GlucoseReading>([]));
    final glucoseReadings = use(glucoseReadingsCapsule);
    final query = (GlucoseReadingQuery q) => q.inTimeRange(timeFrame.toTimeRange());

    use.effect(() {
      final sub = glucoseReadings.watchChanges(query).listen((diff) {
        final updated = buffer
            .addAll(diff.inserted)
            .addAll(diff.updated)
            .removeWhere((gr) => diff.deleted.contains(gr.id))
            .toIOrderedSet();
        setBuffer(updated);
      });
      return sub.cancel;
    }, [timeFrame]);

    use.effect(() {
      glucoseReadings.all(query).then((results) => setBuffer(buffer.addAll(results)));
      return () {};
    }, [timeFrame]);

    return buffer;
  }
```

**Reply by GregoryConrad** [reply]
Took me a second to parse what was going on in that example, but I'd likely make the glucoseReadings capsule not use stream based APIs as a start (unless that's out of your control, such as a bluetooth API).

If you can't change the API, here's the next best thing:
1. Capsule for the current `query` function
2. Capsule for the `glucoseReadings.all(query)` (should this not be a stream too?--confused as to what is going on there)
3. Capsule for the `glucoseReadings.watchChanges(query)`
4. Capsule to combine the previous two capsules together based on whatever logic you need (I'm guessing the `all(query)` returns a future that overrides/trumps all the changes from `watchChanges`)

Step 3 will require a bit of care. My guess is you will need a use.previous or maybe use.data on the current buffer to be able to apply the current diff to it, and return the new buffer with that current diff. The diff part of it (instead of fulls states) makes it a bit tricky. Maybe something like:
```dart
final buffer = use.data(myInitialBuffer);
use.effect(() {
  buffer.value = buffer.value
            .addAll(diff.inserted)
            .addAll(diff.updated)
            .removeWhere((gr) => diff.deleted.contains(gr.id))
            .toIOrderedSet();
}, [diff]);
```

**Reply by opsb** [reply]
glucoseReadings is implemented on top of Electric SQL, `glucoseReadings.all(...)` does a query against the db, `glucoseReadings.watchChanges` receives any changes that are made to the db after the timeFrame is updated (triggered by either apple health updates or changes on the backend which are synced over a websocket). The reason diffs are used is that the dataset can become quite large so fetching the full set on each change becomes quite expensive (the original RxDart implementation ran into this issue before switching to diffs). The tricky bit is step 4, because the changes also include deleted items it's not possible to simply merge the two sets (from `all(query)` and `watchChanges`). 

The changes you suggest make a lot of sense in terms of making the code more idiomatic wrt Rearch. What I'm still not clear on is how to track when side effects have settled within a test. One approach might be to convert a value into a stream.

```dart
const bufferStream = (CapsuleHandle use) {
  const streamController = StreamController<IOrderedSet<GlucoseReading>>();
  const buffered = use(glucoseReadingsBuffer);
  use.effect(() {
    streamController.add(buffered);
    return () {};
  }, [buffered]);
  return streamController.stream;
}
```

Maybe a helper along the lines of 
```dart
expectLater(container, glucoseReadingsBuffer, equals(IOrderedSet(GlucoseReading(...))));
```

Another approach might be https://api.flutter.dev/flutter/package-fake_async_fake_async/FakeAsync-class.html which gives you control over microtasks. The advantage there being that the capsules don't need to expose any information about their side effects to tests.

I appreciate you taking the time to work through this discussion. Hopefully it will be useful for other people as well though.

**Reply by GregoryConrad** [reply]
> What I'm still not clear on is how to track when side effects have settled within a test.

typically, async capsules that you need to test should return a Future, Stream, or an AsyncValue, all of which can be tested fairly easily (using language builtins for when returning Future/Stream, or container.warmUp for AsyncValue). I’d recommend making intermediary capsules that always return a Future/Stream for this case since they’re easier to test, and then a wrapper capsule that use.future or use.stream, often for use in the final widget (assuming you want that state cached).

with the approach I gave, the intermediary capsules would expose future/streams, which you could then verify with the associated dart test package utils

**Reply by opsb** [reply]
I had a go at applying your suggestion

```dart
    GlucoseReadingQuery Function(GlucoseReadingQuery grq) queryCapsule(CapsuleHandle use) {
      final (timeFrame, _) = use(timeFrameCapsule);
      return (GlucoseReadingQuery q) => q.inTimeRange(timeFrame.toTimeRange());
    }

    Future<Iterable<GlucoseReading>> fetchedGlucoseReadingsInCurrentTimeFrame(CapsuleHandle use) {
      final query = use(queryCapsule);
      final glucoseReadings = use(glucoseReadingsCapsule);
      return use.memo(() => glucoseReadings.all(query), [query]);
    }

    Stream<RecordChangeNotification<GlucoseReading, GlucoseReading, String>>
        watchedGlucoseReadingsInCurrentTimeFrame(CapsuleHandle use) {
      final query = use(queryCapsule);
      final glucoseReadings = use(glucoseReadingsCapsule);
      return glucoseReadings.watchChanges(query);
    }

    IOrderedSet<GlucoseReading> combinedGlucoseReadings(CapsuleHandle use) {
      final buffer = use.data(IOrderedSet(<GlucoseReading>[]));
      final fetched = use.future(use(fetchedGlucoseReadingsInCurrentTimeFrame));
      final diff = use.stream(use(watchedGlucoseReadingsInCurrentTimeFrame));

      use.effect(() {
        diff.map((data) {
          buffer.value = buffer.value
              .addAll(data.inserted)
              .addAll(data.updated)
              .removeWhere((gr) => data.deleted.contains(gr.id))
              .toIOrderedSet();
        });
      }, [diff]);

      use.effect(() {
        fetched.map((data) {
          buffer.value = buffer.value.addAll(data);
        });
      }, [fetched]);

      return buffer.value;
    }
```
Is this the sort of thing you had in mind?

**Reply by GregoryConrad** [reply]
Yup! You can likely remove the use.memo in the fetchedGlucoseReadingsInCurrentTimeFrame capsule but other than that it looks pretty good. Now you can await the future/stream capsules and once those are done, re-read the combined capsule and it should have the updated state.

**Reply by opsb** [reply]
Ok great, I understand your thinking. I'm still not sure how I'd go about testing this though, somehow you need to know when the side effects have settled.

**Reply by GregoryConrad** [reply]
@opsb  you should be able to:
1. Read the combined capsule
2. Read and await the future/stream in the two async intermediary capsules in your test
3. Re-read the combined capsule with the updated data

**Reply by opsb** [reply]
In this case I'd regard stream and future providers as private (not because they're exposing fx but just because they shouldn't be part of the public api) so I wouldn't want to rely on those when verifying the behaviour of the capsule. Given the similarities with Riverpod I thought I'd see how they approach testing. This was the best example I found https://codewithandrea.com/articles/unit-test-async-notifier-riverpod/. In that case they're relying on the equivalent of container.onNextUpdate. It's not quite as elegant as waiting for effects to settle but it does seem pretty reasonable.

**Reply by GregoryConrad** [reply]
Ahh, in that case, maybe you can do something like:
```dart
// in your test:
final completer = Completer<()>(); // or whatever it is
container.listen((use) {
  final numBuilds = use.data(0)++;
  final expected = switch(numBuilds) {
    case 0 => const AsyncLoading<int>(),
    case 1 => const AsyncData<int>(Some(321)),
    _ => throw UnsupportedError('produced more times than expected'),
  };
  expect(use(combinedCapsule), equals(expected));
  if (numBuilds == 123) completer.complete(());
});
await completer.future;
```

That should be fine.
